File: objc/Observer/Observer/AppleInc.m
#import "AppleInc.h"

@implementation AppleInc

-(instancetype)init {
    if(self = [super init]) {
        _observers = [[NSMutableArray alloc] init];
    }
    return self;
}

-(instancetype)initWithFaceValue:(double)faceValue {
    if(self = [super init]) {
        _observers = [[NSMutableArray alloc] init];
        _faceValue = faceValue;
    }
    return self;
}

-(void)setFaceValue:(double)faceValue {
    _faceValue = faceValue;
    [self notifyObservers];
}

-(void)registerObserver:(id<Observer>)observer {
    [_observers addObject:observer];
}

-(void)removeObserver:(id<Observer>)observer {
    [_observers removeObject:observer];
}

-(void)notifyObservers {
    for(id<Observer> ob in _observers) {
        NSLog(@"Notifying Observers on change in face value");
        [ob update:_faceValue];
    }
}

@end

File: swift/Observer/Observer/AppleInc.swift
import Foundation

class AppleInc : Observable {
    var observers: [Observer] = [Observer]()
    var faceValue: Double = 0.0
    
    init() {
        
    }
    
    init(_ faceValue: Double) {
        self.faceValue = faceValue
    }
    
    func setFaceValue(_ faceValue: Double) -> Void {
        self.faceValue = faceValue
        notifyObservers()
    }
    
    func registerObserver(_ observer: Observer) -> Void {
        self.observers.append(observer)
    }
    
    func removeObserver(_ observer: Observer) -> Void {
        // Need to implement
        //self.observers.remove(at: self.observers.index(where: { $0 == observer}))
    }
    
    func notifyObservers() {
        for observer in observers {
            print("Notifying Observers on change in face value")
            observer.update(faceValue)
        }
    }
}
